### DISCUSSION TITLE
Observations from 2 games played under 2.7

### DISCUSSION THREAD

**Discussion by Michael2589** [induktio/thinker#24] [General]
Hi Induktio;

The first game was on the new "very huge" map with vanilla rules and mainly vanilla thinker.ini settings (including base spacing 3). I used your options for social_ai and faction_placement.  I played to 2258 - I felt the very huge map was just too huge for my taste :).

The second game was on the new "large" (50x100) map with base spacing 4, a starting expansion_limit of 20, which I increased over the course of the game, and a number of custom changes to alphax (mostly pretty minor; I'll mention them if I think them likely to impact my comments). I used vanilla options for social_ai and faction_placement, and the Drones landed in the Jungle and took out me (but it was so early that I got a restart), the University, and the Spartans

Observations

SECRET PROJECTS

An issue on the extra huge map was that the AI's just let me take all of the early game SP's. They seemed to be too busy ICS'ing to even try to compete with me for the SP's.
On the large map I split the first 8 SP's: 4 for me, 4 for the Drones. The Drones landed in the Jungle, had a very strong start, and have been the strongest AI faction throughout the game, so that's not obviously a problem.
As a simple rule, I think that when the AI gets to six bases and is at peace it should check that it's built recycling tanks and rec commons at its capital and then just start building a SP at its headquarters.

GARRISONS

A really big problem on the large map was that by 2270 (turn 170) the base garrisoning in 2.7 was really, REALLY bad. (I haven't played much since 2.0 so I can't tell whether the changes mentioned in the changelog for 2.7 made things better or worse. But it's really bad as of 2.7.) About a third of the land bases of AI Hive and Sparta had no land units in them. This is doubly bad because both factions have SE settings that give them police bonuses, which they're not using. I can send you a save file if that would be useful.

I think it's partly a problem with the build choice AI, too. The Hive have built a lot of formers (20 land formers for 12 bases) although I'm not sure whether that's actually a problem. More problematically, they've been building a lot of aircraft at the expense of not building cheap defensive infantry, which means that now that I've landed on their continent I'm going to walk all over them. On the other side of Planet, the Drones are simultaneously doing the same thing to the Spartans. 

If I were telling the AI what to do:
"Defensive infantry" are infantry with higher defence than offence, except before the discovery of synthmetal armour, when 1-1-1 infantry count as defensive infantry.

Production AI:
If they don't have a defensive infantry unit in the base, then that becomes their current build choice. They can switch to it instantly without penalty because they're the AI. (IIRC they do this in the vanilla game when you kill the last unit in their base.)
If they have researched Non-Lethal Methods, and their Police SE setting is -1 or higher, and the base has a population of <pick a number> or higher, then build a police defensive infantry as part of the garrison.

Movement AI:
First, if a defensive infantry is in a base, calculate target garrison size for that base:
Base of 1 defensive infantry per base.
Add 1 to the target garrison size if the faction has Police setting of 1 or higher (so allowing at least 2 police units per base) and Support setting of 1 or higher (so supporting at least 3 free units per base). [The Hive running Police State and the Spartans running Power are examples of factions that will often meet these criteria.] Maybe this only kicks in if the base is size 4 [pick a number] or higher.
Add 1 more if the game difficulty is at Thinker (difficulty 5) aka the cost_factor from thinker.ini is 8 or less. [Trading off some of the AI bonuses against the reality that the AI can't predict where attacks are going to come, so heavily garrison everything.] Again, this might only kick in for larger bases.
... for a maximum target garrison size goal of 3.

If the unit is a defensive infantry unit in a base, and the number of defensive infantry units in that base is less than or equal to the garrison goal, then the defensive infantry shouldn't move.
If the unit is a police infantry, and the base's Police setting is -1 or higher, don't move unless there are at least two police infantry units in this base.

Yes, this means they'd need to build a LOT of defensive infantry :). They should have enough defensive infantry for these garrisons and more to participate in the offensive armies. But the thing is, the game encourages these factions to build these infantry. These factions typically have high Support values, allowing them several free units, and Police ratings which means the garrisoned infantry give an economic benefit even when the faction's at peace, and, realistically, it's good for the AI to just stash defensive infantry at every base because the AI is not good at assessing the strategic situation and guessing where attacks on it are going to be made.

In my wish list of things the AI could do, it would be aware of whether it's (1) at peace; (2) nominally at war but no units are being destroyed, e.g. in a war with someone on the other side of the map; (3) in a hot war; (4) in a war and losing bases. Easier said than done.

BASE FACILITIES

The buildings that I build a lot of and the AI hardly builds any of are hologram theatres, research hospitals, tree farms, and to a lesser extent hab complexes.

For example, the Drones and the Peacekeepers are doing well in my second game. The Drones have 5 size 16 bases and 9 bases that are size 13 or larger, and they have a grand total of 1 hologram theatre. (I'm using 4-square base spacing, I modded the pop limit for bases with hab complexes to 16, and the Drones started in the Jungle, so lots of big bases.) The Peacekeepers have a size 18 base and they are just getting around to building their first holo theatre in it now.

Vanilla AI behaviour is that if a base goes into riots, and it has no rec commons, the production immediately switches to rec commons with no production penalty. If the base has rec commons, but no holo theatre (assume they don't have the Virtual World), it immediately switches to building a holo theatre. They may do the same with Research Hospitals as well; IIRC the vanilla AI builds a lot of research hospitals if it gets that far. I think this behaviour would be worth replicating.

I also noticed that the AI was building few Net Nodes (some of the AI's have built no net nodes). IMO all the AI factions should build a few early so that they can cash any alien artifacts they've found.

A longstanding issue with the terraforming algorithm is that the AI rarely builds solar panels on tiles it's built farms on. There are many tiles in AI territory with just farms, no second feature. If they don't have the technology or secret project to build condensors, building a solar panel on tiles with a farm should be standard. Sensors should be built on forests, not farms, because there's no opportunity cost to building a sensor on a forested tile.

**Comment by induktio** [comment]
Okay interesting. That's a detailed writeup. There's some things that really rised my eyebrows though. You didn't mention difficulty settings? That's pretty important as far as the difficulty pacing goes. If you set expansion_limit at 20, that's pretty low for Thinker AI's unless you've simultaneously enabled expansion_autoscale. It's mostly there to put some kind of an upper ceiling so the AI doesn't get too close to the unit limits.

You mention the base garrisoning being somewhat broken, but it's hard to correlate that kind of behaviour with the stuff that I've seen on the test games. Usually the AIs don't have any shortage of defensive infranty units. Maybe sometimes they have too many obsolete units, but having a bunch of bases with no garrison units should not normally happen. For example the last garrison unit should NOT exit the base square if there's no other defenders. It feels like there's something else going on with custom game settings, maybe alphax.txt changes.

If all the defenders are destroyed, the AI should switch production to garrison units. It's basically the very first item that is checked in the build queue. They also have routines to build extra facilities on rioting bases, but hologram theaters are built more rarely because of the maintenance cost. Also network nodes shouldn't be that low priority in the queue, unless the bases produce so little energy they get skipped because of the cost-benefit check. So the AI was building something, but it was not a) garrison units, b) additional base facilities?

Many of the things you mentioned are already implemented in the decision making, for example "(1) at peace; (2) nominally at war but no units are being destroyed, e.g. in a war with someone on the other side of the map; (3) in a hot war; (4) in a war and losing bases." E.g. Thinker has special priorities if they start losing a lot of bases and switch production accordingly. I'm just confused as to why these didn't seem to have effect on your game. It could be useful if you can send the config files and some save games at various points. The address is admin at induktio.net.

**Comment by Michael2589** [comment]
I am going to put together two zip files with save games/thinker.ini/alphax.txt, but it'll probably take until tomorrow morning, so I'll post this for now.

The something that the AI is building that's neither garrison units nor additional base facilities is needlejets. The (tweaked) Drones and the (vanilla) Hive in particular really, really like to build needlejets.

I am seeing the AI's switch production to garrison units when the last defender is destroyed, so that's not the problem. The problem that I'm seeing is them moving the last garrison unit out of bases, and if that base is building a needlejet that'll take 10 turns to finish, it keeps on building that needlejet.  (Which makes sense: why would you reconsider production every time a unit leaves a base?)

Game settings: both games: Librarian (4) difficulty. Random tech with revised_tech_cost=0, eco_damage_fix=1, ignore_reactor_power=1.
Very huge map game used social_ai=1, expansion_limit=50, expansion_autoscale=1, base_spacing=3. rare_supply_pods=1, faction_placement=1, Thinker 2.7 alphax file, factions Gaians (player), Usurpers, University, Morgan, Sparta, Caretakers, Drones. This is the game where the AI's didn't build any SP's in the opening game. All of the other observations I made were from the second game.
Large map game used social_ai = 0, expansion_limit started at 20 and I've been adjusting it upwards as the game's gone on, expansion_autoscale=1, social_ai=0, base_spacing=4, rare_supply_pods=0, faction_placement=0, hybridized modded alphax.txt, modded Drones (Industry +1 instead of +2, Planet -2 instead of Research -2, agenda Planned instead of Eudaimonic), modded Gaians (Impunity: Green but it didn't matter since I didn't run Green until after I finished the Cloning Vats), Gaians (me), Hive, University, Morgan, Sparta, modified Drones, Peacekeepers.
Because faction_placement=0, the Drones, Gaians, University, and Spartans all started on the same continent. The Drones started in the Jungle and eliminated me so early that I respawned. They eventually drove both the University and the Spartans off the continent into the sea. Because I had Spoils of War on, the Drones got a strong early tech boost from conquering both me and the University, so it's not surprising that they built 4 SP's before any of the other AI's had built any. Despite this early tech boost, and the Drones starting in the Jungle, and that I had modded the Drones so that they had no research penalty, the Peacekeepers eventually overtook the Drones in tech and built the final two SP's built by the AI. The most iconic moment so far has been when the Free Drones conquered *ahem* liberated Sparta Command and built a Punishment Sphere there (the only Punishment Sphere that has been built by anyone during the game)!

Most of the changes I've modded in have been about moving SMAX stuff into the late game so that the early game plays more like vanilla SMAC, which is why I'm not expecting most of the modifications to make much difference. But there are two changes in particular, both only in the second game, that are making a difference:
Base spacing set to 4, which is resulting in larger bases, and I think that's what has caused the lack of holotheatres to become an issue
I disabled any tech from allowing the construction of boreholes, and since I got the Weather Paradigm that means no one is building boreholes in the second game, and THAT means that the AI's energy production per base is very low, since they're not getting much energy from their condensers and mines, and you say that would cause them to think that network nodes aren't worth building, which is what I'm seeing.

**Reply by induktio** [reply]
Remember you should have autosaves from every turn unless they already got overwritten by playing the turn again (hopefully not). As a general rule, you can't except the AI to be very challenging if it's playing humans at even odds or worse. I feel like most of the older players use the two highest levels anyway. The lack of SP building probably reflects the lack of productivity because of cost_factor. On the higher levels I haven't seen any reports that AI would have problems building SPs. Maybe I should just change the default to cost_factor=12,11,10,9,8,7 to bring it in line with the other settings.

Generally I'm not too much into building solar panels since they seem to be just wasting former turns compared to other options, but disabling boreholes might have just that kind of ripple effect elsewhere which causes a lack of energy production. Also no energy means no drone alleviation from psych allocation. And no credit rushing. So it could be a big problem. Drone management is probably one of most tricky areas to code for the AI and I would almost favor some solutions that simplify the game mechanics there rather than code a bunch of additional edge cases for the AI. Also normally you would probably see the AI building much more Spheres but if it built only one it kinda suggests something is wrong with the productivity.

Most games are probably won by acquiring needlejets early and rushing opponents, so it's hard to say what could be the problem of building many of them. Are they not used effectively? You seem to want the AI to build a) more facilities, b) more secret projects, c) more garrisons. But at the same time it's hampered by the cost_factor and lack of energy, so it can't do all of these at the same time and manage to fend off invasions without having needlejets. So basically it needs more production bonuses.

Sometimes the AI might need to build some extra solar panels if e.g. boreholes are disabled but most of the time forests are a much better option. Are forests still 1-2-1? If the cost benefit calculus for terrain enhancements changes drastically, it might cripple the AIs since it does cost benefit checks there only partially (fungus vs. forests for example). Hard to say without seeing the whole situation. Just guessing.

**Comment by Michael2589** [comment]
[Thinker 2.7 ungarrisoned bases.zip](https://github.com/induktio/thinker/files/7612490/Thinker.2.7.ungarrisoned.bases.zip)

Hopefully that worked.

In the year 2270 save game, the following are examples of bases with no land unit garrison that are not building land units: 
Hive: Great Collective, Manufacturers Warrens, Hole of Aspiration, Laborers Throng, Paradise Swarming, Factory Drift, Port Yang
Sparta: Hero's Waypoint
Drones: University Base, Edistvo-Unity (as I mentioned, the University had a bad game)
They are all building units (as opposed to base facilities), but they're not building ground combat units.
Also look at all the medics producing psych in those size 16 bases. The Drones have the net income to pay for holo theatres, and a holo theatre only has to free up one specialist medic to work as a specialist technician to pay for its maintenance, and then whatever the second worker that's no longer a drone produces is entirely profit. In these cases it would probably free up a third medic, since the holo theatre would also increase the productivity of the medics it doesn't free up.  For my own faction, I regard holotheatres as just something I'm going to need to build if I'm spacing my bases four tiles apart. The expense is a necessary one in that situation which it wouldn't necessarily be the case if one was using a spacing of 3.
The year 2300 save game shows the Drones with their one punishment sphere.
One caveat is that many of the base facilities in Morgan's bases in 2270 and in Yang's bases in 2300 were built by me (after I took their bases, defeated them, built facilities in their bases, and returned their bases), so e.g. AI Yang hasn't been building tree farms, I built the tree farms in his bases.

All terraforming options are vanilla except for the house rule against building boreholes. To be clear, I'm not trying to analyze whether the AI should build forests or farms + solar panels. I'm observing that the AI is building farms, and not building anything else on the tile until it has the tech for condensors. I'm claiming that, if you've decided to build a farm rather than to build a forest, then also building solar panels is a better option than leaving it at just a farm. If you're going to work that tile, then the worker is going to consume two food, so the net profit of a tile is its production - 2 food. So a rainy rolling tile with a farm will produce for example 3 food (after restrictions lifted) and 1 mineral, for a net production of 1 food and 1 mineral, but if it also has a solar panel it'll have a net production of 1 food, 1 mineral, and 1 or more energy, which is a notable productivity improvement for the rest of the game. Before nutrient restrictions are lifted it'd have a net production of 1 mineral and 1 energy vs. 1 mineral.
The way I play, not having boreholes doesn't particularly matter because I'm producing energy from solar panels, tidal harnesses, and forests. I build over 95% tidal harnesses compared to mining platforms. But the AI isn't building solar panels and it builds few tidal harnesses, so being also blocked from building boreholes becomes a big deal. Having said that, building boreholes isn't a possibility for most AI's in the early game, and not building solar panels or many tidal harnesses slows down its research before it researches the tech to build boreholes, which leaves it at a large disadvantage even after it's researched that tech.

Reviewing old save games of the game where I thought it didn't build early SP's, it did try, just too little and too late, and only got a couple of SP's that I didn't particularly want and weren't much use to the factions that built them. I agree that playing on a higher difficulty level would probably have made it competitive.

The needlejets were being used effectively on a tactical level, but strategically they were being wasted. The AI was not using them to fend off an invasion: they were using them to try to soften me up before they launched their amphibious invasion against me. (My invasion happened because after they used all those resources for their aerial attack, I noticed that they had little in the way of defences.) And if it had been just the Drones (a strong AI) vs Morgan (a weak AI), that would have been a good strategy, and the AI needs to be able to do that. But it wasn't a good idea for the Hive, with an economy much smaller than of mine, to decide to spend most of its resources building needlejets, forfeiting all its future economic growth, in order to fight a war of attrition with me. Though the prequel to that had been that I had taken over three Hive seabases that were in exposed positions, so I suppose they had good reason to think themselves in an existential war. But I had no intention of following that up with a land war against the Hive before they made a nuisance of themselves. And then there's Morgan, who is much weaker than the Drones and is using aerial units to attempt to wear them down prior to making an amphibious invasion, and because Morgan is in a Pact with me and I'm wearing the Drones down it's quite possible that that his invasion will capture a base. How you make the AI understand these different strategic situations I don't know.

Acquiring needlejets early and rushing opponents is, I think, more of a player thing to do than an AI thing to do. Unless things are very unbalanced there's only a narrow window where one side has needlejets and the other doesn't. In a game where the player is working towards a Transcend victory, the AI is going to spend many turns when it has access to needlejets but does not have the economic/technological advantages to use them for a rush.

The way I see combat in this game is that you have cheap defensive units and expensive offensive units. (True whether the offensive units are rovers or needlejets.) The offensive units will almost always kill a defender when they attack (because if the odds weren't in their favour they wouldn't have attacked). So you win if either (1) the enemy exposes one of the expensive attackers killing one of your cheap defenders, and then you kill the exposed attacker, which is a good trade or (2) you find an opportunity to attack an expensive attacker when its not covered by a cheap defender, which is even better. Therefore you need a mix of offensive and defensive units, because the higher proportion of defensive units the more units you will have and the more attacks you can withstand, but too many defensive units and you can't make many attacks. The question then becomes: what is the best ratio?
I think that the Hive and the Spartans in the 2270 save are building too many expensive units given the mix of units they currently have. Neither of them has enough resources to win their wars, regardless of what they do, but I think they'd sell their virtual lives more dearly with more defensive units, and the Hive had enough resources that if they'd turtled earlier I'd have left them alone. You can look at the save and see what you think.  (It doesn't help the Spartans that the vanilla AI for choosing an SE has decided that they should run Knowledge values and the Drones are mind controlling their bases, but that's on me for choosing that vanilla AI :)

also EDIT: this started off with me seeing the AI moving their garrisons out not knowing that that's a bug if it's happening. Maybe the saves will help you see the conditions where it happens. I'll try to find before and after saves showing it happening later this weekend when I get time. /EDIT

Just to clarify, the SP's issue was specifically about the early game situation when you're playing with faction_placement=1, or if the vanilla faction placement has put the AI far enough away from everyone else that they're too far away from each other to fight a war in the early game. I think you're correct that on a higher difficulty they would have built a reasonable number of early SP's.

**Comment by Michael2589** [comment]
I realized that my thinking about the Hive was being influenced by the fact that they were the Hive running police state, in particular. So +2 police +2 support means that they support 4 free units per base and can use three of them as police, so building two cheap infantry units for every base makes economic sense for the Hive in particular, costing no support but giving them economic benefits with drone management, but not for factions in general.

I was thinking that adding a difficulty "game rule" that gives the AI's with Police ratings of -4 or worse a +1 situational modifier to Police could work. So -5 becomes -4, -4 becomes -3, and -3 or better isn't changed. That way SE choices with Police penalties would still carry a price that the player could see, but it would be much easier for the AI to manage, and it would be the same sort of thing as quietly giving the AI +1 industry +1 growth when the player is in 1st place.

**Comment by induktio** [comment]
If you're now playing some additional games you should upgrade to v2.8. It made some changes that could affect some of the stuff here. However some issues I will address separately in a new develop build soon.

- Garrisoning needs some additional checks to make sure a land defender unit is present in the base at all times. In some cases it could consider air combat units as defenders, which is a problem.
- The lack of energy from boreholes exposes some problems there. AI former priorities need some updates to build more solar.
- Related to same point, more holo theatres need to be built. Another small change. There were some checks that did not work well with some game situations.

Those are probably the most important changes. Of the other things I'm not sure what can be turned into actual changes in the AI code.

**Reply by Michael2589** [reply]
That covers 90% of my concerns, with the other 10% being more feature requests than bug fixes :).

**Comment by induktio** [comment]
Develop build 20211129 is now available from downloads and it also provides this new feature:

- To ease calculations, base hurry dialog will now display the minimum required hurry cost to complete the item on the next turn. This assumes the mineral surplus does not decrease, so take it into account if adjusting the workers. When entering a partial payment, this minimal amount will be the default choice in the dialog, instead of the full hurry cost like previously.
- Cost factor is updated to have a smoother progression for each difficulty setting.
- Holo theatres are built more often in the bases when they are needed.
- Formers will build solar panels slightly more often on suitable rainy/moist+farm tiles.
- The whole terraforming code is adjusted to do more cost benefit checks to account for more possible alphax.txt rule changes.

It's extremely time consuming to tune the terraforming code though. There's just so many edge cases and different settings. This version should manage situations without boreholes available among other things better than previously. If you find some situations where the formers make weird decisions, let me know.

Garrisoning code should be updated to ignore air units for defending purposes but I think the lack of defenders previously was mostly related to the cost_factor setting. At one time I actually tried playing the mod with cost_factor=5 and it was maybe doable but the endless unit spam made conquering bases a real struggle. It was common to see border bases stacked with 8+ units in that case.

**Comment by DrazharLn** [comment]
Thinker-doer has terraforming AI that is intended to handle modded
alphax.txts, have you compared your solution to theirs? Might be something
useful that could be done as a synthesis

I haven't looked at their code, but they talk about it as if it's doing
some kind of search of the possibility space.

**Comment by induktio** [comment]
Well, if you look at WTP's former code, it's a completely separate solution that does not integrate on Thinker's codebase at all. It's somewhat hard to read such code and it is quite long. WTP's code might also rely on some additional patched features on the game. I think it patched condensers for example. So I haven't seen yet any evidence it would ge generally better than Thinker's version. Some player commented the former code was broken there, so go figure. If you know some specific area in the code to compare for improvements that might be more helpful. Otherwise Thinker's version should be quite complete by now except it needs some additional testing for different rulesets.

Those new changes in Thinker should now handle any mods such as setting borehole yield at 0-4-4 like WTP did or even disabling them altogether. Though I'm not going to change it for now since the mod is mostly using vanilla settings. Forests can be modded also but I don't recommend it, since any change there could alter balance too much because the values only use small integers and fractions are not possible.

**Comment by induktio** [comment]
Speaking of forest tiles, there might still be some situations where the current solution does not work well, especially if forest yield is modded lower than the default. It still needs some changes to account for all situations with landmarks etc. Another thing to look for is if it still builds too many condensers. They can easily bring too much nutrients for the bases.

**Reply by Michael2589** [reply]
While my personal mod highlights the issue, 6 factions every game will be playing without boreholes until they research Ecological Engineering (a tier 4 tech) and the boreholes won't produce a lot of energy until the faction researches Environmental Economics. So the AI needs to know how to play without boreholes - it's the usual situation during the early game.

**Reply by induktio** [reply]
For the most part they know how to play the early game without them, they manage to research the tech later anyway? ;)

For sea tiles, I'll probably just keep it at half solar/half mining. It seems to be good enough for most situations. On land, solar tiles just doesn't seem very competitive because of extra former turns compared to forests, and they might get replaced by condensors later. Maybe at higher altitudes solar should be used more. One energy from forests combined with rivers here and there seems to be a more flexible approach. That is if forests produce at least one energy, otherwise they would not be very competitive. I have to adjust the former logic a little bit but it won't be that big of a change compared to current version hopefully.

**Reply by induktio** [reply]
New develop build is now available and it adds these things:

- Config option natives_weak_until_turn: Native units spawned by the planet fight at half strength until this turn is reached. Also in WTP with a different name.
- Retool exemption doubled to 20 minerals.
- Extensive reworking on former improvement priorities.

Basically the main idea is to never build something on the tile that doesn't increase the total resource output. It's easier said than done because the game doesn't expose any API for this, but this version implements full cost-benefit checks for forest/fungus/borehole tiles. That means if Thinker selects one of those options, it should always be the item that increases production the most.

Additionally farm/solar/mine tiles should never be replaced if the replacement does not increase the total output. It should also use drill aquifer command slightly more often on better locations. It's very significant for increasing energy production overall. This logic should manage various game situations/configs quite well and I'll probably keep it like that with the exception of fixing bugs.

**Comment by induktio** [comment]
Uploaded a new develop build again. Changes in this version:

- New config options: neural_amplifier_bonus, dream_twister_bonus, fungal_tower_bonus.
- Update tech_cost_factor to the same progression than cost_factor.
- Design_units creates police garrisons more often unless AAA is available.
- Additional fixes to former movement code mostly.

**Comment by Michael2589** [comment]
I started a new game with the November 29 build at difficulty 5, resetting conquer_priority to 50 about 100 turns in.  Base spacing was 4 which will result in some differences from your standard games.  Overall, the AI did really well - the strongest AI's had tech parity with me as late as 2240, before decisively falling behind after that.  The hologram theatre issue seems to be resolved now.  (I still think it should build more of base facilities and fewer units in general, but that's a base facilities issue and not specific to holo theatres.

However, the AI is still abandoning bases.

I looked into four instances that I particularly noticed (because I was in wars with those AIs).
Two cases were associated with transport foils: in one case, there was a transport foil in a sea base with three infantry units, then the next turn the base was abandoned and shortly afterwards I killed a foil nearby that was carrying two infantry and found a third infantry wandering around the island that the sea base was adjacent to. In the second case, one turn a base had one infantry and was about to complete a transport foil; the next turn the base was empty and the base was supporting a transport foil that was presumably the one it had completed the previous turn.

In one case, the base had been reduced to -1 food and negative net minerals, (for some reason it was supporting 7 formers among other units). Not surprisingly, that base was not able to build a new garrison.

In the fourth case, I was advancing on an AI base from both its northeast and its southeast, and all three infantry in the base, one offensive and two defensive, left the base to try to put up an forward defence, one unit per tile, leaving the base garrisoned with one probe team and two formers. Unwise, because staying in the base would have given them better odds, and also being a player I was able to use mobile units advancing on roads supported by a copter to attack the now-undefended base from the south and west.

I can provide save games if that'd be helpful.

**Reply by induktio** [reply]
Sometimes a bunch of units gets assigned to one base for support when the engine readjusts things. Not necessarily a bad thing because then it's less support for other bases. Sometimes the base might run out of surplus minerals because so many units get reassigned there though.

So in general, do you think the base garrison issue got any better after these recent changes? Is it a significant issue for the AI factions in the big picture?

**Reply by Michael2589** [reply]
It's hard to say whether the base garrison issue got better or not.  (I think it did, but that's a very subjective impression.) They're doing better this game than last game, but the AI doesn't have air units yet, which was what was giving them a lot of problems last time, and I'm playing on difficulty 5 instead of difficulty 4, which means they have more of everything.  So I've got multiple different input variables and it's not obvious what the influence of each is on the AI's behavior.

In the big picture?  It makes them less efficient (it means that factions like the Hive that are intended to have the ability to use police instead of base facilities don't make as effective use of that ability as they might), but this is a complex game and there are always going to be things making the AI less efficient. They're still substantially more efficient than the vanilla AI. It occasionally allows me to snipe bases that I shouldn't have been able to snipe, though more likely it'll cause a base that could have held out for two turns to fall in one, but that's unlikely to change the winner of a war, and mistakes like this are why we give the AI production advantages.

Perhaps more importantly, it's a mistake the player will see and which looks very bizarre from a player's point of view, and it's a mistake the vanilla AI doesn't make, which is easier for the vanilla AI because it isn't trying to accomplish as much as this AI does.

Obviously I think it's important, but I'm just one player.  Do you talk with other people who use the mod?  Is there a Discord or something?

It occurs to me, that if your AI uses unit orders similar to the ones in the player's interface (e.g. "go to", "hold", and all the various terraforming orders), you might be able to create a new command, "guard", and then AI functions that are trying, for example, to organize multi-unit attacks, could know that they can grab units with "hold" orders but just ignore units with "guard" orders.  Or maybe that's not how it works at all.

**Reply by induktio** [reply]
I'm trying to gauge how often this happens with current code, so that affects what kind of fixes it needs. It's very important for bug reports to be reproducible. So it would be helpful if you can construct some kind of a savegame where this situation becomes apparent. It doesn't even have to be a normal game, you can also use Ctrl+K editor functions to alter the units or even create some AI-vs-AI game.

There's a discord for SMAC which I can recommend, however it's maintained by someone else. The link is also shown next to r/alphacentauri toolbar: https://discord.gg/yPVhhvWBBF

**Reply by Michael2589** [reply]
[AI abandoning bases example.zip](https://github.com/induktio/thinker/files/7721611/AI.abandoning.bases.example.zip)
Here's a save file which was played under the December 10 patch. The 2288f file is the end of year/turn 2288, and Autosave_2289 is the beginning of the next turn. At the end of 2288 the Hive have land units in Sparta Command, Bunker 118, Society Grid, Blast Rifle Crag, and Discipline Tubes, and they move all land units out of those bases during their turn.

The game was played on the huge map of Planet at difficulty setting 5 and with base spacing of 4.

If you look more closely at the save, (for example by examining the garrisons tab of the F4/Base Reports screen for the Hive), you can see that the Hive have concentrated a huge force of land units at Fecundity Tower, and many of their air units at the Hive, their capital base, which is nearby. The history is that they have been at vendetta with the Morgans and the Drones for a while, and only recently restarted their vendetta with me.  The Morgans and the Drones are to the Hive's east, and Fecundity Tower is their most easterly base, while I am to the Hive's south.  So the Hive are moving their army east to attack the Morgans/Drones.  Unfortunately, the Hive are on a different continent from the Morgans and the Drones, so their army is just stationed on their eastern shore looking across the straits at the Morgan territory on the other side of the sea.

Something else these save files illustrate is the importance of the AI routinely building tree farms. Look at all of the size three Drone bases in the middle of huge forests, and while I point at Drone bases, University Base headquarters itself falls into this category as well.  Some of those bases have a lot of infrastructure (it is turn 188) and if they had tree farms they'd quickly grow to size 7 and the tree farms would quickly pay for themselves.  And, even though I've banned the AI from building boreholes, the AI has still managed to cause sea level rise from all of the ecodamage they've done.  Every tree farm the AI builds is one less fungal bloom the AI causes and has to deal with.
